File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
      if(s.length()<t.length()||s==null||t==null)return "";  
      HashMap <Character,Integer> need=new HashMap<>();
      for(char c:t.toCharArray()){
        need.put(c,need.getOrDefault(c,0)+1);
      }
      HashMap<Character,Integer> window=new HashMap<>();
      int r=0,l=0,strt=0,valid=0,min=Integer.MAX_VALUE;
      while(r<s.length()){
        char ch=s.charAt(r);
        r++;
        if(need.containsKey(ch)){
            window.put(ch,window.getOrDefault(ch,0)+1);
            if(window.get(ch).intValue()==need.get(ch).intValue()){
                valid++;
            }
        }
        while(valid==need.size()){
            if(r-l<min){
                min=r-l;
                strt=l;
            }
            char re=s.charAt(l);
            l++;
            if(need.containsKey(re)){
                if(window.get(re).intValue()==need.get(re).intValue()){
                    valid--;
                }
                window.put(re,window.get(re)-1);
            }
        }
      }
      return min==Integer.MAX_VALUE?"":s.substring(strt,min+strt);
    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        # i=0
        # j=0
        # res=""
        # while j<len(t) and i<len(s):
        #     res+=s[i]

        #     if s[i]==t[j]:
        #         j+=1
        #     i+=1
        need={}
        for i in t:
            need[i]=need.get(i,0)+1
        left=0
        formed=0
        req=len(need)
        start=0
        window={}
        minVal=float('inf')
        for right in range(len(s)):
            window[s[right]]=window.get(s[right],0)+1
            if s[right] in need and window[s[right]]==need[s[right]]:
                formed+=1
            while formed==req:
                    if (right-left)+1<minVal:
                        minVal=(right-left)+1
                        start=left
                    if s[left] in need and need[s[left]]==window[s[left]]:
                        formed-=1
                    window[s[left]]-=1
                    left+=1
    
        return "" if minVal==float('inf') else s[start:start+minVal]